### ISSUE TITLE
Xbox One Wireless Adapter for PC Support

### ISSUE THREAD

**Issue by mejamiewilson** [360Controller/360Controller#156]
Microsoft have released the wireless dongle for Xbox One controllers (I don't know if it's compatible with 360 controllers). Support for this guy would be great, as you can have 8 controllers connected to it. 

I have bought one if anyone wants to see any info from it.

**Comment by FranticRain** [comment]
We are definitely going to need to product and vendor IDs for that thing. Also, it doesn't support 360 controllers. Just Xbox One controllers.

**Comment by mejamiewilson** [comment]
Here's the result of running `system_profiler -detailLevel full` on Mac OS X.

Good to know that it doesn't support 360. Is the scope of this project locked down to 360 controllers? Or is it ok to have Xbox One specific functionality?

```
XBOX ACC:

          Product ID: 0x02e6
          Vendor ID: 0x045e  (Microsoft Corporation)
          Version: 1.00
          Serial Number: 310241
          Speed: Up to 480 Mb/sec
          Manufacturer: Microsoft Inc.
          Location ID: 0x14200000 / 7
          Current Available (mA): 1000
          Current Required (mA): 500
          Extra Operating Current (mA): 0```

**Comment by FranticRain** [comment]
Xbox One is welcome, I'm the guy that does all of the Xbox One work for this driver.

**Comment by mejamiewilson** [comment]
Fantastic! If this issue gets to the top of the pile, let me know what I can do to help. Driver related stuff is definitely over my head.

**Comment by geowar1** [comment]
FranticRain: I have a lab full of Macs, an Xbox One controller and the new wireless USB dongle for Xbox One controllers if you want/need me to test anything. You may contact me at "geowar 1 @ mac . com" (without spaces).

**Comment by natbro** [comment]
i haven't yet been able to get through the XBox One dongle's initial handshake to talk through as works with XBox One wired. anybody else have ideas/leads? @FranticRain - ditto, happy to send you a dongle and/or controller if you don't have yet: natbro-at-gmail-dot-com.

**Comment by geowar1** [comment]
My thought was to use a beagle (USB packet sniffer) on a windows PC and record USB traffic to the dongle 1st sans the controller and then with.

**Comment by FranticRain** [comment]
You really don't even need the beagle. You can just do it with Wireshark and USBPcap, but a beagle would probably work best. If you're interested in the Wireshark method, [you can check it out here. These are instructions for sniffing the Xbox One controller itself.] (http://www.definedbehavior.com/reverse-engineering-the-xbox-one-controller/) Once you figure out the packets, its basically just an exercise in figuring out what means what and then you're good to go. Or if you don't want to figure it all out, just send me the info and I'll try and piece it together for you. I can't guarantee it'll work though. I've never looked at the 360 wireless adapter code before.

**Comment by geowar1** [comment]
FranticRain has already got the 360 dongle working on OS X. We need to do the same for the Xbox One. (I'll assume that's what you meant).

**Comment by FranticRain** [comment]
No, I was a little vague there. I would need to extend the 360 wireless adapter code to make it work with the Xbox One code. I've just never looked at the code this driver uses to make wireless adapters work. That's what I meant.

**Comment by electric-monk** [comment]
Ahem, I did the original wireless 360 dongle support ;P

I just used the Windows driver + some USB snooper app (no idea what it was at this point, but its icon may have been Snoopy). This didn't give me any functionality the Windows driver didn't have though (and I did use a Beagle for some things, such as to deduce wired ChatPad support though.

Some people with more XBox 360s to lose than me used a Beagle on the wireless dongle inside the real console to work out the off command (and perhaps Wireless chatpad support, I think someone sent the required magic code to me to make that work) as the 360 basically just contains the same thing that the dongle does to speak to the controllers. To get XBone wireless support working nicely may also require someone to disassemble a real console and Beagle it, too.

**Comment by FranticRain** [comment]
My bet is that Windows 10 is tied close enough to the Xbox One controller that disassembly will not be required. With stuff like the Xbox app, and the app to rebind controller buttons as part of Win 10, we should be able to get everything we need.

**Comment by natbro** [comment]
yes, seems likely that crazy disassembly won't be required. but hopefully no irritating signed handshake or other nonsense :( i'll try to capture the new Win10<->new-dongle handshake in USBPcap on Monday and share, my home Windows machine is currently in a sad state, can't do tonight.

**Comment by natbro** [comment]
too busy today to cap, hopefully asap this week.

**Comment by oxycoder** [comment]
Any update on this progress?

**Comment by marzman** [comment]
I have a mac running windows 10 on a virtual machine -- I have successfully connected the adapter to my windows 10 but have not been able to get my xbox one controller to pair with it -- any ideas on how to get this working ? thanks for any help

**Comment by FranticRain** [comment]
@electric-monk We have a sniff of the packets going to and coming from the wireless adapter with USBPcap. You can find it over here: https://github.com/360Controller/360Controller/issues/160#issuecomment-159952216
The file works with Wireshark.

**Comment by electric-monk** [comment]
I'll take a look when I find a bit of time ;)

**Comment by CleanCut** [comment]
@electric-monk @FranticRain I'm willing to contribute [a dongle](http://www.amazon.com/Microsoft-Xbox-Wireless-Adapter-Windows-one/dp/B00ZB7W4QU/ref=sr_1_1?ie=UTF8&qid=1453241001&sr=8-1&keywords=xbox+one+wireless) if that would help get wireless xbox one controllers working.

On a side note, when I plug my brand-new Xbox One controller via the USB cable, it gets detected, but the keys are very wrong.  I was able to mostly remap the keys, except for the analog sticks.  Is there some trick to getting the analog sticks remapped correctly?

**Comment by RodrigoCard** [comment]
please check this version @CleanCut , the mappings are fixed https://drive.google.com/file/d/0B5PLCKoMd4zGNy15NFR1TTRxVkE/view?usp=sharing

**Comment by CleanCut** [comment]
@RodrigoCard That fixed the mappings, but there's other problems.  In order to not hijack this thread I created issue #207 to discuss it.

**Comment by Queuecumber** [comment]
Are you guys having any luck with this? We aren't having much success on the Linux side

**Comment by FranticRain** [comment]
No one has been working on it as far as I know. I haven't bothered to buy a wireless adapter yet, and @electric-monk hasn't said much of anything.

**Comment by drejjmit** [comment]
Tried SnoopyPro as electric-monk suggested but nothing shows up. No data at all.
[](http://i.imgur.com/nlxm6bq.png)

**Comment by FranticRain** [comment]
USBPcap + Wireshark is a way better way of capturing data. We already have a sniff though, it really just needs to be tested and implemented. From #160, https://drive.google.com/file/d/0B381VMfcesrlNWZwRkpoTXJBaDA/view?usp=sharing

**Comment by drejjmit** [comment]
Did 2 caps with USBPcap and 1 with Wireshare. Maybe they will help
[https://drive.google.com/file/d/0B3-WQPnZBRZhRmlNZjltU1pWTms/view?usp=sharing](https://drive.google.com/file/d/0B3-WQPnZBRZhRmlNZjltU1pWTms/view?usp=sharing)

**Comment by FranticRain** [comment]
Finally got a wireless adapter. Will take a shot at it later this week and over spring break. Word on the street is that it is encrypted. Sounds like this is going to be fun.

**Comment by TyanColte** [comment]
Good Luck

**Comment by maxiedaniels** [comment]
I just bought the adapter and wanted to see how the progress on this is going? Thanks!

**Comment by FranticRain** [comment]
Going nowhere fast. I'm by no means well informed with the USB spec, but that whole thing is a complete mess. I can't just fake it until I make it like I did with the wired controller. I doubt I'll ever figure it out, and I'm pretty busy with school anyway.

**Comment by stevenvachon** [comment]
I just bought a wireless adaptor for my Xbox One controllers as well. Disappointed that I have to use the included USB cables. Anxious for a new release!

**Comment by haiduc32** [comment]
So if I was to invest some sleepless nights trying to write the driver, can anyone give me a crash course into the libraries and at what should I be looking?
@FranticRain are you positive it's encrypted? any links?

**Comment by Queuecumber** [comment]
@haiduc32 As far as I know we have no idea if it's encrypted or not, it might just be complicated

**Comment by FranticRain** [comment]
@haiduc32 For driver documentation, its basically nonexistent. Any that exists is poorly written and outdated. The best documentation you can get on IOKit drivers is [here.](https://developer.apple.com/library/mac/documentation/DeviceDrivers/Conceptual/IOKitFundamentals/Introduction/Introduction.html#//apple_ref/doc/uid/TP0000011-CH204-TPXREF101) Other than that, your best bet is to just study the wireless driver code currently in this driver. That is how I wrote the Xbox One wired driver. 

As for documentation on the actual packets, I recommend [my own documentation](franticrain.github.io) because, as far as I know, its the only documentation. It has information on how to capture packets from devices, the format of the Xbox One wired packet, and some various stuff I have discovered about the wireless packets.

Now, as for USB packet documentation. The driver *probably* follows the USB HID guidelines, but I can't honestly prove that. For that documentation, you are going to need to look at [the USB website.](http://www.usb.org/developers/hidpage/)

I wish you the best of luck, and if you have any questions, let me know.

(I just pushed the wireless documentation to my website, it usually takes a couple minutes to update the github.io page for no apparent reason.)

**Comment by haiduc32** [comment]
Thanks, I was more interested in getting a fast understanding of the project libraries, how it works for the 360 wireless adapter so I could try to get a fast start at the One. Starting the driver from scratch doesn't sound very tempting..

**Comment by FranticRain** [comment]
The only documentation I know about the wireless 360 controller adapter is the original Tattiebogle stuff. http://tattiebogle.net/index.php/ProjectRoot/Xbox360Controller/WirelessUsbInfo

**Comment by haiduc32** [comment]
Thanks. Will try to sniff it tonight, see if I get any patterns. I'm concerned about the encryption, as it would be a show stopper. Will keep you posted if I make any progress.

**Comment by haiduc32** [comment]
HELP! So far my experience was like going to a bar, getting wasted, then getting on a motorcycle and driving blind folded up hill through the woods. 

I got to the point where I think I know what I should be sending on the bus, but I have no idea what I'm actually sending and where.. so.. **does anyone know how to intercept USB traffic on OSX El Capitan?** It's ridiculous but I haven't found any tools..

**Comment by FranticRain** [comment]
Wow, you know a lot more about these packets than I do! Awesome!

As for intercepting traffic on El Cap, there is no good way, but if you are working with the driver, the driver should let you know every time its gotten data. You can use IOLog to output that data into whatever format you want. It will show up in your "system.log" in the Console.app. I recommend adding a prefix to all of your data so you can quickly do a search in the log for your prefix and filter out all non-essential packets. That's what I'd do at least. I don't know if anyone has any better suggestions.

**Comment by haiduc32** [comment]
@FranticRain thanks, but I did that already. I need to check exactly how the packets are formatted, especially the headers. I am trying to get a vm up with 10.10, people have reported that USB interception works on that version.

**Comment by FranticRain** [comment]
@haiduc32 Good luck then. I'm super excited to hear you have a plan!

**Comment by Queuecumber** [comment]
@haiduc32 Hey, I have a small request, when you get a chance how hard would it be for you to document some of the thing you found out w.r.t. the protocol for the wireless driver? I'm working with the Linux side of things are we are having a lot of trouble as well.

**Comment by haiduc32** [comment]
@Queuecumber I'm not a "documentation" person.. I think code will be the best documentation. But let's see first if I manage to get it working, or let's have a chat if you want to try things in parallel. Drop me a private message if you want to share contact info (skype etc). 

@FranticRain thanks..

**Comment by tomlube** [comment]
Anyway I can help with the wireless support?

**Comment by haiduc32** [comment]
@tomlube if you have 2 controllers, I'd appreciate a sniff of the USB traffic.

**Comment by FranticRain** [comment]
Turns out I only had 3 AA batteries, so in my sniffs, Controller 1 was powered from the wall, while Controller 2 was running batteries. Hope that's not a problem. I captured on Windows 8.1 using a 2015 controller (Controller 1) and a 2013 controller (Controller 2).

twoControllerInitial - Activate sniff, plug in adapter, Windows does some USB driver installing garbage like it always does, two previously paired controllers connect.
twoControllerIdle - Just an idle of the two connected controllers, started sniffing after connected
twoControllerAPress - I pressed A on Controller 1 and then A on Controller 2.

Note that these controller numbers may not be accurate, they may be swapped. I only know what I believe to be 1 and 2 since there is no LED ring.

File was zipped using 7-Zip on Windows 8.1, should be perfectly safe: [Two Controller Sniffs.zip](https://github.com/360Controller/360Controller/files/212103/Two.Controller.Sniffs.zip)

**Comment by geowar1** [comment]
Use the HID Manager to register a Get Report callback and then (hex) dump everything that it receives.

**Comment by haiduc32** [comment]
hey guys, wanted to drop by and say hello.. also I managed to get the wireless adapter into pairing mode (led blinking) but the SoB is stubborn and doesn't want to pair with anything.. any way, I got 2000 lines of code just to get to this stage, and my eyes hurt.. I'll have time only next weekend to continue and will release some code once I manage to pair with the controller so other people can try it out if it works with their adapters.

**Comment by haiduc32** [comment]
hi guys, I need a few more sniffs, as for pairing I need a specially constructed packet, which is different in 3rd party scans and I still haven't figured out exactly how to build it.
so far I got sniffs from FranticRain and drejjmit, more sniffs especially while pairing, will be highly appreciated!

**Comment by haiduc32** [comment]
hi guys, I need a few more sniffs, as for pairing I need a specially constructed packet, which is different in 3rd party scans and I still haven't figured out exactly how to build it.
so far I got sniffs from FranticRain and drejjmit, more sniffs especially while pairing, will be highly appreciated!

**Comment by haiduc32** [comment]
hi guys, I need a few more sniffs, as for pairing I need a specially constructed packet, which is different in 3rd party scans and I still haven't figured out exactly how to build it.
so far I got sniffs from FranticRain and drejjmit, more sniffs especially while pairing, will be highly appreciated!

**Comment by tomlube** [comment]
Hey Haiduc, how could I sniff these for you? Do you need multiple controllers/adapters/Windows/OSX/what are the requirements and how could I help?

**Comment by FranticRain** [comment]
@tomlube This should get you on the right track for how to sniff. Sniffs need to be on Windows. OS X ones won't help.

**Comment by looterz** [comment]
@haiduc32 Here is a sniff I ran on a USB Hub with only a keyboard and the wireless adapter connected. It contains packets of the adapter searching for a device, finding the device and pairing, then about 5-10 seconds of traffic after a successful pair.

**OS**: Windows 10 x64
**Devices**: XBox Wireless Adapter for Windows (v21.50.45.656), XBox One Controller (v6.2.11059.0)
**pcap**: https://db.tt/36tawx5e

**Comment by haiduc32** [comment]
Thansk, I've seen the linux thread and plan to use their sniffs for checking as well. I am kind of busy with life at the moment and not sure how much time will be able to afford on spending on this project, so things might go slower. I'll try to post my current work so if anyone is available, to be able to chip in.

**Comment by tomlube** [comment]
OS: Windows 10 Pro x64
Devices Xbox One Wireless adapter for Windows/Xbox One Controller
USBPcap: http://www35.zippyshare.com/v/Lz32B1fS/file.html

**Comment by rpcastagna** [comment]
Hey @haiduc32 if you start a repo/branch with the code you have so far I might be able to help get it a little further; I've been pretty busy with life stuff too but I spent long enough pulling these packets apart that I feel obligated to try to make some meaningful contribution out of it, haha. I'll post the captures I'd made while playing with it, also.

**Comment by rpcastagna** [comment]
Various captures: 
OS: Windows 10 Student x64
https://www.dropbox.com/s/fwvqdj3cdz7hdcj/Xbox%20One%20Wireless%20Caps.zip?dl=0

**Comment by haiduc32** [comment]
Hi, so this is my fork: https://github.com/haiduc32/360Controller
I'd appreciate if someone could try it out (I mean developers that know how to run it) and check if their wireless adapters are entering into pairing mode (led blinking). It does for me.
For those that want/have time to contribute to the effort, the class is OneWirelessGamingReceiver. There are checkpoint comments with line numbers that correspond to sniffs from dejjmit: https://drive.google.com/file/d/0B3-WQPnZBRZhRmlNZjltU1pWTms/view?usp=sharing the 1.pcap
I started a doc for my investigation: https://drive.google.com/file/d/0B3-WQPnZBRZhRmlNZjltU1pWTms/view?usp=sharing if anyone wants to contribute let me know and I'll provide read&write permissions.

**Comment by FranticRain** [comment]
I went ahead and made a signed build of this so people can try it out without having to go through the trouble of building it. [You can get that here.](https://drive.google.com/file/d/0B5PLCKoMd4zGbmNtVTdENVN0ZzQ/view?usp=sharing) **PLEASE NOTE THAT THIS WILL NOT MAKE YOUR WIRELESS CONTROLLER WORK AND IS HIGHLY UNSTABLE. IF YOU COMPLAIN ABOUT THIS NOT WORKING I WILL COME TO YOUR HOUSE AND SNAP ALL OF YOUR XBOX ONE CONTROLLERS IN HALF,**

I got my wireless adapter into pairing mode by pressing the button on the adapter. This caused the light to blink. I then took a 2013 controller and put it into a sync mode. 10-20 seconds later, I received a kernel panic. I didn't expect the controller to pair, but I definitely didn't expect a kernel panic.

**Comment by haiduc32** [comment]
@FranticRain thanks for trying it out. don't know why it panicked, did not happen for me. maybe highlight the **RISK OF KERNEL PANIC!!!** so people are trying it on their own risk.

**Comment by FranticRain** [comment]
@haiduc32 It's important to note that I still kernel panic whether or not I activate the pairing mode. Based on your console outputs, its doing an awful lot of idle talking, and then it panics. I can't get anything specific out of it, but I definitely won't be messing around with it for a couple weeks. Hooray for finals.

**Comment by tomlube** [comment]
Have tried it out and I can confirm that in indeed works to start pairing and then kernel panics.

**Comment by haiduc32** [comment]
@FranticRain it's not idle talking, just Microsoft overthinking it..

@FranticRain and @tomlube, can you look into the panic report what is the panic reason?

based on the fact that I can't reproduce it, I can only guess that it's because it depends on some controller/adapter Id that I missed/ignored. I can't debug it on my side so if any dev is willing to investigate I'd be happy to give some pointers, if not ... it will be a question of when and if I get a second set of controller/adapters to try it out. look me up on skype **haiduc32** if you want to help.

I'm going forward with my controller at the moment, trying to get it working, to see if I can get it done. Once it's working I'll focus on making it compatible but until then any help is appreciated.

**Comment by geowar1** [comment]
Try to reproduce it via the guest account and/or a non-development machine.

The development user/environment often allows thing to work (for debugging purposes) that cause a crash in a non-privileged environment.

**Comment by tomlube** [comment]
Can confirm it kernel panics on a guest account.

**Comment by FranticRain** [comment]
Here's a link to all of my panic logs I got while trying to debug the issue: https://drive.google.com/file/d/0B5PLCKoMd4zGckdoNjItNExkVjA/view?usp=sharing

**Comment by tomlube** [comment]
https://mega.nz/#!Vwo1VZpT!30pF-p9aqw_iHH-P0lkQjCPn3zZL4P-Iya77hQpsy5Q there are mine

**Comment by tomlube** [comment]
Any progress at all with this? I see there is 8 commits since this build has been released :)

**Comment by haiduc32** [comment]
hey guys, I'ver been pretty busy with other projects, but now I'm refocusing on this one, so fingers crossed. can't promise anything for the next two weeks as I have to figure out where i got stuck and continue from there. 

in other news, you're probably all aware that a new xbox one gamepad has been announced and that although i suspect they use the same hardware, it's very possible they will be updating the communication protocols and that might be rolled out even to older xbox one gamepads through updates. that probably means trouble.

**Comment by stevenvachon** [comment]
Ugh, the new Xbox One controllers are bluetooth-based 😕

**Comment by natbro** [comment]
Oh really? Where did you find that out? Actually that could make life a lot
easier depending on *how* it's using Bluetooth.

**Comment by WalrusCode** [comment]
I think they mentioned it in their press conference.
And even then they may update the comm protocols as haiduc32 has said which is going to be a bitch.
Regardless the whole Bluetooth thing would make it a lot more accessible to people, buuuut I bought a freaking adapter so I'm pissed xD.

**Comment by stevenvachon** [comment]
Yes, I'm annoyed that I just bought new controllers and an adapter three months ago.

**Comment by WalrusCode** [comment]
Well the new controllers will apparently need to work with older Xboxes so they should still work.
I'm literally going to use mine as much as possible.
Microsoft is always golden with timing. NOT!

**Comment by FranticRain** [comment]
Please keep in mind that Microsoft's implementation will always get better performance and lower latency than Bluetooth. Bluetooth is a usable option, but it is inferior to Microsoft's implementation. More discussion about the Bluetooth implementation of this driver can be found in #291

**Comment by haiduc32** [comment]
I got my controller sending button press messages with the wireless dongle. before you start celebrating there are a few issues that are still pending:
1. looks like it works only for me.. probably some codes were device specific. don't know yet how to figure it out (3rd party scans don't help).
2. currently it works with a controller and a wireless dongle that have been paired previously on windows. I don't know how to unpair my controller to test with a brand new pairing.
3. if you remove the adapter and plug it back in the same usb port, it doesn't work at all - unless a more advanced os x dev can help me out on that, it'll be that way..
4. now I have to figure out what code needs to be added to make it actually work with games.. help from devs is welcomed.
5. sometimes the adapter does not initialize correctly. no idea why

all the code is checked in into my fork. it's ugly but it works (at least for me).

**Comment by FranticRain** [comment]
I'll see what I can do regarding 4, but I busted my wireless adapter trying to tear it open (whoops) so I can't help with anything else unless I decide to buy another one, which I'd really rather not.

**Comment by haiduc32** [comment]
I did open the adapter but stopped at the RF shield. if you went further mind sharing a few high res photos of the board? maybe try to read the markings off the chips and post it? i'd be curious to know what they used.

**Comment by FranticRain** [comment]
The chip in the middle of the previously shielded area is marked as such:
```
MEDIATEK
WT7600UAN
1537-BMAL
ETP3CY23
```
Those white rectangle things from the top are marked "HX NX HX NX"
Nothing else under the shield really has any markings. You'll notice where I ripped off two resistors just below the bottom right corner of the Mediatek chip, as well as the busted sync button.

![img_20160709_103942451](https://cloud.githubusercontent.com/assets/4530595/16708697/09a5827a-45c2-11e6-868b-0dbe9b608c55.jpg)
![img_20160709_103953648_hdr](https://cloud.githubusercontent.com/assets/4530595/16708699/0c42a9b8-45c2-11e6-9a0f-367e66a5b783.jpg)
![img_20160709_104016694_hdr](https://cloud.githubusercontent.com/assets/4530595/16708701/0e628ccc-45c2-11e6-9270-ad52ae9b681b.jpg)


And yes, I did take these pictures on top of an original Xbox for no apparent reason other than it was convenient at the time.

I also quickly realized that my plan for helping you out really wasn't going to work particularly well if I couldn't test it myself, so I'll tell you what my thought process was. The stuff you are looking for is going to be `receivedUpdate` and `receivedHIDUpdate` in `WirelessHIDDevice`. I don't know how you are going to need to change them to work with the Xbox One packet order, but `XboxOneControllerClass::convertFromXboxOne` will probably be of help.

**Comment by Candunc** [comment]
Hello! First off, thank you for the work on this driver, I've used the wired Xbox One controller implementation in the past and it has worked flawlessly. I came back today to install it again on my Macbook and was curious about the wireless implementation. It looks like you've put a lot of effort into it, but you are still not there.

Anyways, I was curious if there was anything else I could do to help. Do you need more USB captures from various configurations, and has anyone figured out if it is possible to dump the USB capture straight from the Xbox?

**Comment by eligao** [comment]
Nice work!
FYI, once I connect the controller to Windows 10 using wire, I will need to do manual pairing again to get it working wireless. Not sure if this is a brand new pairing though.

**Comment by mejamiewilson** [comment]
@FranticRain Ping me an email (it's on my github profile page) if you'd like a new wireless adapter. No pressure on working on this (it's secondary to other wired issues) but I'm happy to send one your way.

**Comment by jmfranz** [comment]
Hey folks, any news on this issue?
If someone (in canada or us) needs a dongle to work let me know :)

**Comment by jaspertandy** [comment]
If this needs hardware throwing at it, I'm happy to donate some. Devs linked to the GH, post an amazon wishlist or something and I'll send pads/dongles/whatever's needed.

**Comment by stevenvachon** [comment]
If I'd only bought my controllers 3 months later I coulda got Bluetooth 😢

**Comment by renarsvilnis** [comment]
Bumping @jmfranz question ` any news on this issue?`

**Comment by dannyharding10** [comment]
Any update on this? This would be amazing!!

**Comment by stevenvachon** [comment]
Is there a place that we can donate for this feature?

**Comment by kylewlacy** [comment]
I noticed this issue sort of lost steam over the last 2 years or so, so I set up a bounty through BountySource here: https://www.bountysource.com/issues/27658018-xbox-one-wireless-adapter-for-pc-support

Hopefully this is enough to get someone to tackle this issue head-on finally! A few resources I've seen that may be useful to anyone looking to take this on:

- The conversations in this thread as well in the sister issues at paroj/xpad#14 and paroj/xpad#25 have included lots of discussions and dissections of the Wireless Adapter and the protocol
- From @FranticRain I found an analysis of a packet dump from the Wireless Adapter: https://vestigl.com/packetformat/wirelesspacketformat.html
- Some people have provided various packet dumps from their Wireless Adapters: [[1]](https://github.com/360Controller/360Controller/issues/156#issuecomment-183472059)[[2]](https://github.com/paroj/xpad/issues/25#issuecomment-258638658)[[3]](https://github.com/paroj/xpad/issues/14#issuecomment-269816473)[~[4]~](https://github.com/360Controller/360Controller/issues/156#issuecomment-215904210)[~[5]~](https://github.com/paroj/xpad/issues/14#issuecomment-184694503)[~[6]~](https://github.com/360Controller/360Controller/issues/156#issuecomment-183488520)[~[7]~](https://github.com/paroj/xpad/issues/14#issuecomment-152785727)[~[8]~](https://github.com/paroj/xpad/issues/14#issuecomment-173383092)
- @haiduc32 had a fork of 360Controller that seemed to get the Wireless Adapter into pairing mode, as well as a document with another analysis of the packet dumps (mentioned earlier up in this issue): https://github.com/360Controller/360Controller/issues/156#issuecomment-216017112
- @Adubbz gave a really good summary of the known technical information about the adapter here: https://github.com/paroj/xpad/issues/14#issuecomment-318029036. The most important one being that the Wireless Adapter is based on the MediaTek MT7216US NIC, but the device has no persistent storage. So. their theory is that the firmware is written to the device when it is initialized. Luckily, there is already an open-source firmware for the MT76xx family here: https://github.com/openwrt/mt76/tree/master/firmware (which should be compatible with the Wireless Adapter's chipset?)

**Comment by Two-Tone** [comment]
Holy hell, that's a huge bounty.

**Comment by FranticRain** [comment]
In light of this, I'm going to share some concerns I have over cracking the encryption on the wireless adapter. This doesn't appear to just be a byte packet that we are cracking, but rather an encryption scheme. I'm all for bringing support to devices on other platforms, but cracking encryption may very well put us on Microsoft's bad side. With an exposed protocol, we open the door for knock-off controllers in a way that Microsoft was clearly trying to avoid. I would rather not put this repository in jeopardy, if at all possible. Leaving the wireless code closed source may be the best option. I don't know for sure, that's just my two cents on the subject.

**Comment by stevenvachon** [comment]
Microsoft has since moved to Bluetooth, so I don't think they would care as much anymore.

**Comment by atar-axis** [comment]
https://github.com/atar-axis/xpadneo/wiki/Microsoft-Xbox-One-Wireless-Adapter

**Comment by rgov** [comment]
@FranticRain (I am not a lawyer.) First, it's not necessarily the case that there is any need to "crack" the encryption. I did not see any reference to cryptography in the linked analyses and dumps. If there were, from my experience with encrypted protocols (but not this one), I would assume that the adapter and controller perform a handshake that results in a shared secret. Re-implementing this handshake algorithm would not constitute "breaking" the cipher.

Second, there is no restriction against reverse engineering and re-implementing a wireless protocol or cryptographic algorithm in the United States (where both Microsoft and GitHub are located). The Digital Millennium Copyright Act specifically provides safeguards for reverse engineering for "interoperability" purposes, as well as for analyzing cryptographic methods.

Third, there is a long history of Microsoft's cryptography being broken, from LM hashes to [broken Xbox crypto](https://www.zdnet.com/article/new-xbox-security-cracked-by-linux-fans/) to [poorly designed wireless keyboards](http://travisgoodspeed.blogspot.com/2011/02/promiscuity-is-nrf24l01s-duty.html). Security researchers routinely publish works about reverse-engineered Windows internals, etc. Yet I do not know of any instance where Microsoft has threatened any developer or project over this content.

**Comment by rubenjavier** [comment]
@FranticRain just wanten to chime in and say that there is already Xbox One wireless controllers knockoffs for v1 controllers (no bluetooth) and they work with the Xbox One and with the wireless adapters.
There is also knockoff receivers adapters beign sold, both of this (controllers and adapters) has been on the market for months now... (I've got one of this controllers and it works)

Here is one of this controllers reviewed in youtube:
https://www.youtube.com/watch?v=JKmXE2UmGxs

Here is the link to one of aliexpress vendors, with controllers in black,white and the receiver, in case the link stop working, just look in aliexpress for "xbox one wireless controller"

https://es.aliexpress.com/store/product/For-Xbox-One-Wireless-Remote-Controller-Jogos-Mando-Controle-For-Xbox-One-PC-Gamepad-Joypad-Game/1092020_32814926791.html?spm=a219c.search0604.3.54.249e41755OOzNK&ws_ab_test=searchweb0_0,searchweb201602_2_10065_10068_10547_10546_10059_10884_10548_10887_10696_100031_309_10084_10083_10103_10618_10307_10820_10303_525_449,searchweb201603_60,ppcSwitch_3&algo_expid=885723d4-a2da-4e19-a9fb-414fd0626a29-10&algo_pvid=885723d4-a2da-4e19-a9fb-414fd0626a29&priceBeautifyAB=0

**Comment by go0ro0** [comment]
$15 added to bounty

**Comment by FranticRain** [comment]
Just adding a little reminder here that as of 10.15, this driver will most likely be deprecated due to the introduction of Apple's kext notarization scheme. Long and short of the issue is that drivers will have to be built with Xcode 10 or later and then validated by Apple in order to be used. The main problem with that is that this driver is still built against Xcode 6.4. It is unclear whether we will be able to build against an outdated SDK with Xcode 10 or later since that isn't officially supported. And then it is unclear if we can still get notarization using that scheme. The solution is a full re-write of the driver, but the new USB code Apple added in 10.11 has a memory leak in it, so I have not upgraded the driver and have no intention to until the memory leak is resolved. Therefore, if someone DOES implement this support, they will likely also be on the hook for re-writing the driver.

**Comment by go0ro0** [comment]
https://gist.github.com/devkid/4b3bd50760504d1b93ea684cfd3ed895

May be of interest...

**Comment by go0ro0** [comment]
I personally am hoping for Linux and eventual android support, I'd like to use the dongle on the Nvidia Shield.

**Comment by medusalix** [comment]
I have good news for you guys 😃. During the past few weeks I implemented a custom driver for the Xbox Wireless Dongle based on [`libusb`](https://libusb.info/). Although it is currently *Windows only*, the basic functionality could be ported to macOS quite easily. Here's a short GIF to show you that it's already working very well:

![demo](https://user-images.githubusercontent.com/8124898/63446607-cf811000-c43a-11e9-9b02-50a5a36679fe.gif)

There's still a few things I haven't figured out, such as the automatic channel selection. This is why I need **_YOUR HELP_**: If you have the newer and smaller *`Xbox Wireless Adapter for Windows 10`* or the older and clunkier version of the dongle, you can help me by providing an up-to-date [`USBPcap`](https://desowin.org/usbpcap/) capture of the communication between your PC and the dongle. The data has to be captured from start to finish, so please start the recording BEFORE you plug in the dongle and stop the recording after you have done some actual button presses on your controller.

Once I'll have collected enough data I will provide a *testing* version of my Windows driver so you can try it out on your own PCs. Maybe we will finally have an open driver for the wireless dongle 😄.

**Comment by rubenjavier** [comment]
@medusalix that is just awesome, sadly I just have the bigger older version, I'll try to get someone with the newer version to help or borrow one when I get back from vacations. Tweeted this to libretro, batocera, retropie and recalbox in hopes of finding more users with that dongle

@twinaphex Could you please tweet this info to try to get help from the libretro users? thanks in advance

**Comment by medusalix** [comment]
@rubenjavier Thanks for tweeting, that might get the attention of a few more users. I'll need captures for both the old and the new dongle, so I'm interested in your capture as well. According to my experimentation the old dongle uses the Mediatek `MT7612` while the new one is based on the `MT7632`.

**Comment by go0ro0** [comment]
Just read this, I have 3 of the new smaller dongles, need instructions to make the capture and where to send upload it.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by go0ro0** [comment]
I also have the chatpad so I can capture with and without, would a capture with headset plugged into the controller be any use also?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by medusalix** [comment]
Instructions for making the capture:

- Install [Wireshark](https://www.wireshark.org/). Make sure to tick the option to install **USBPcap** in the installer.
- At the bottom of the Wireshark window you should see multiple USBPcap entries.
- You have to identify which one is the correct one by clicking on one of them to start the capture.
- Now plug in the dongle. If you see lots of messages appearing, then it's probably the correct one. Otherwise you have to repeat the procedure with one of the other USBPcap entries.
- Turn on the controller and wait for it to pair (light stays on). Press any random buttons on the controller.
- Stop the capture and save the captured data. Upload the created *.pcapng file to any file hoster you like and post the link.

Thank you.

**Comment by ugly95** [comment]
@medusalix 
You might want to get in touch with the developer of the xpadneo driver: [https://github.com/atar-axis/xpadneo](https://github.com/atar-axis/xpadneo).  The developer has done a great job getting the Bluetooth version of the XBox One S controller working on Linux.  I believe there was some desire to get the wireless adapter working as well.

**Comment by medusalix** [comment]
@ugly95 That sounds like a good idea 😃. My top priority would be a driver for macOS, especially because the bounty is targeting 360Controller, but I'm more than willing to provide a Linux driver too. I'm just not entirely sure if this project is even maintained anymore (see [comment](https://github.com/360Controller/360Controller/issues/143#issuecomment-506508536) by @FranticRain). Maybe a first implementation with xpadneo would be a good idea? @atar-axis

**Comment by FranticRain** [comment]
@medusalix You'd be correct in your assumption that 360Controller currently does not have an active maintainer. If you'd like to get a request merged in, ping me because I no longer follow the repository. However, I also won't be signing anything anymore, so you'll have to get your own certificate to get it actually released. So 360Controller will probably be the most difficult version to actually get released. Sorry about that.

But I've gone ahead and gotten a sniff for you, as I know getting these can be a huge pain in the butt. Everyone starts out excited about it and then either gives you a crappy quality sniff or doesn't at all. But I got you one on a clean port. Xbox One S-type controller. New dongle. Sniff includes dongle connection, then I hit A, B, X, Y, LB, and then RB. Let me know if you want other controller types with the same dongle, as I have every iteration of the mainline Xbox One controller just sitting around here.

It's marked as a "txt" because GitHub is a well made and intelligently managed service, but I promise you it's a pcap. :P
[FranticRain.txt](https://github.com/360Controller/360Controller/files/3538713/FranticRain.txt)

**Comment by medusalix** [comment]
@FranticRain Thanks a lot for the nice capture 👍. There were a few minor things I added to my driver, although it didn't differ very much from the captures I already did, which is a good thing (as it means that my driver should also work with somebody else's controller). Though there are still many things I haven't figured out:

- [ ] Automatic channel selection (hard)
- [ ] Controller rumble (should be easy)
- [ ] Auto-off when computer is shut down (need a WiFi capture for that)
- [ ] First-time pairing (can't do this, my controller is already paired)
- [ ] Support for multiple controllers (should be easy)

I would really appreciate if you could do a capture with multiple controllers connected to one dongle 😃.

**Comment by atar-axis** [comment]
@medusalix wow, i am impressed 😊 and i would be very interested to get in touch with you. The biggest problem in linux always was the missing dongle/chipset support. They added that last year or so. Before that I made some progress with the low level monitoring mode but I never successfully decribted gamepad data.

**Comment by medusalix** [comment]
@atar-axis Your documentation and devkid's [gist](https://gist.github.com/devkid/4b3bd50760504d1b93ea684cfd3ed895#file-xbox-one-wireless-protocol-md) were really helpful! I actually started by investigating how the linux driver works. I added the missing VID & PID and recompiled the kernel driver (or should I rather say the whole kernel because of signature enforcement lol). Then I found out that the driver only supports the MT7612 (probably used in the older dongle) and not the MT7632 that is actually inside the new dongle. I still got basic WiFi functionality working by making the driver think it's a MT7612 (though I got lots of errors in the logs). That's basically why I decided to implement my own driver from scratch.

**Some more information about the protocol**:
These are the only things required to get the controllers to pair:

`Beacon frames` -> `Association request` -> `Association response` -> `Handshake`

You need no probe requests, no probe responses and no QoS null frames. Those things may only be necessary for the initial pairing.
Another interesting thing is that half of the handshake can actually be skipped. The controller starts sending the button actions after a certain point where you can stop responding to the handshake messages.

**Comment by FranticRain** [comment]
@medusalix Encountered a show-stopper bug in my Windows install, so I nuked it, and re-installed. Turns out that makes Windows forget all of the controller pairings. So here's a sync sniff.

[Sync.txt](https://github.com/360Controller/360Controller/files/3557967/Sync.txt)

**Comment by belegdol** [comment]
[xbox_one_controller.pcapng.gz](https://github.com/360Controller/360Controller/files/3561012/xbox_one_controller.pcapng.gz)
I am not sure if this is the correct capture, let me know in case it isn't and I'll try another usbpcap id.

**Comment by medusalix** [comment]
@belegdol Yes, that's it 😃. Thank you very much.

@FranticRain Thanks, with your help I got the pairing figured out 🎉. It's pretty straightforward and simple actually. 

I'm working on the driver for macOS right now, I'll let you guys know when it's ready for testing 💻.

**Comment by atar-axis** [comment]
Hey medusalix, would you mind sharing what you found out? :) Maybe we could usw that for xpadneo too, furthermore I would gladly invite you to collaborate in xpadneo in General :grin:

**Comment by medusalix** [comment]
@atar-axis Sure, I'm actually planning on writing a blog post after I publish the driver, where I summarise all the findings I made and share some protocol details. The driver for `xpadneo` would definitely be the next thing I'd like to work on. Anyway, here's how the pairing works:

The MT7632 contains a section in memory where the beacon frames are stored. You can write *multiple different* beacons to that memory. When the button on the backside of the dongle is pressed or released, the driver creates an additional beacon with a 802.11 frame type of *0111* ([Reserved](https://en.wikipedia.org/wiki/802.11_Frame_Types)). Additionally, the data of the actual beacon frame gets slightly modified: `00, 00, dd, 10, 00, 50, f2, 11, 01, 10, 01, 95, 2c, 99, 00, 00, 00, 00, 00, 00`.
Following that, the dongle receives a response from the controller, still with a frame type of *0111*, the dongle answers with a similar packet.

After those steps, the association flow continues as usual.

**Comment by spphinslove** [comment]
@medusalix I have adapter for windows10 and two controller(one without bluetooth other one with bluetooth and 3.5 mm in-diameter)，and I have installed win10 and manjaro. how can I help you.

**Comment by spphinslove** [comment]
@medusalix 
[xbox.zip](https://github.com/360Controller/360Controller/files/3578017/xbox.zip)
1. I open the wireshark and start cap
2.plug in adapter and press search button
3.turn on the controller (the one with bluetooth) and press search button.then paired success.
4.press any button on controller.
5.use old controller to  repeat 3 and 4.
6.press any button on old controller.
7.stop cap and save file.

**Comment by spphinslove** [comment]
> @belegdol Yes, that's it . Thank you very much.
> 
> @FranticRain Thanks, with your help I got the pairing figured out . It's actually pretty straightforward and simple.
> 
> I'm working on the driver for macOS right now, I'll let you guys know when it's ready for testing .

Any news?

**Comment by medusalix** [comment]
@spphinslove I'm still working on it 😄. Thank you for the capture. USB is causing problems on macOS, I'm unable to do bulk transfer with >512 bytes (maybe I'm missing some kind of auto-fragmentation option). The firmware upload requires the firmware to be sent in large chunks, so that's the kind of thing that prevents me from continuing my work. I might publish the unfinished source code later this week.

**Comment by medusalix** [comment]
![final](https://user-images.githubusercontent.com/8124898/65839605-e30f7880-e30e-11e9-8b52-d369e0cedd40.gif)

@FranticRain @atar-axis @belegdol @spphinslove

**Comment by Two-Tone** [comment]
Is that on a Mac? I ask because the repo is Mac focused but that interface looks like GNOME 3 as there isn't a standard title bar.

**Comment by FranticRain** [comment]
@Two-Tone That's 360Controller on macOS. It's dark mode with orange accent color. He's in the System Preferences menu, which has a different title bar style than most apps. Plus, the camera shot is just so low on the screen that you don't see any menu bar.

**Comment by 13MaxG** [comment]
@medusalix Is the source available?

**Comment by medusalix** [comment]
@13MaxG I will publish the source code soon. The handshake between the dongle and controller is still not 100% reliable, so I'll have to get that fixed before I release it.

**Comment by Two-Tone** [comment]
You don't have to get it 100% reliable before releasing it. Drivers are
worked on out in the open all the time and with this many people and
developers interested in this having more eyes on the code is better.

If you're worried about bad drivers circulating around, just don't provide
an installable binary version. Anyone who is honestly interested in working
on this will be able to compile the code themselves, assuming it doesn't
require some crazy build environment.

**Comment by FranticRain** [comment]
Wireless Xbox One controller support has now been merged into master. I don't know what the process is for claiming bounties, so I'll leave this open until I get further word from @medusalix on what he needs to receive payment.

**Comment by dannyharding10** [comment]
This is so great. What do we do to upgrade?

**Comment by FranticRain** [comment]
You'll need to disable SIP on your machine and build the `modern-rewrite` branch from source using `build.sh`. Then you can run the installer and you should be good to go.

**Comment by medusalix** [comment]
@FranticRain Unfortunately this issue needs to be closed for me to claim the bounty. Do you mind if I open up a new one where people can discuss issues and suggest new features?

**Comment by FranticRain** [comment]
@medusalix By all means!

**Comment by medusalix** [comment]
Please take a look at [this](https://github.com/360Controller/360Controller/pull/50) issue. Same story, rumble wasn't implemented at first. You are of course free to decline the bounty payout, but the `new device` tag is usually used for a first, not necessarily feature-complete implementation.

**Comment by rubenjavier** [comment]
what about the linux implementation? I thought that the biggest use case would be for linux and unix like systems that doesnt have this drive, for distros like retropie, recalbox, batocera or android... suddenly I saw a focus on mac on the last leg of the driver creation, which have a very small gaming community (probably the smallest)

**Comment by FranticRain** [comment]
@rubenjavier This is a macOS repo for a macOS driver. Linux is out of scope, and off-topic.

**Comment by rubenjavier** [comment]
> @rubenjavier This is a macOS repo for a macOS driver. Linux is out of scope, and off-topic.

sorry mate, my bad

**Comment by FranticRain** [comment]
These changes are now released as part of [alpha 4](https://github.com/360Controller/360Controller/releases/tag/v1.0.0-alpha.4)
